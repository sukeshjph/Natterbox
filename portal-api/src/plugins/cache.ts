export default class Cache {
  memoryCache

  constructor(memoryCache) {
    this.memoryCache = memoryCache
  }

  async get(id) {
    const result = await this.memoryCache.get(id)
    return result
  }

  async set(id, token) {
    const result = await this.memoryCache.set(id, token)
    return result
  }
}