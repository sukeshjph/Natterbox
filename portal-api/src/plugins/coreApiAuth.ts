import axios from "axios"

export default class CoreApiAuth {
  gatekeeperBaseUrl

  lassoBaseUrl

  constructor() {
    this.gatekeeperBaseUrl =
      "https://gatekeeper.redmatter-qa01.pub/token/sapien/organisation"
    this.lassoBaseUrl = "http://lasso.qa01.redmatter.com/auth/token/coreapi"
  }

  async getLassoToken(orgId, userId, token, logger) {
    const gatekeeperToken: any = await this.getGatekeeperApiToken(
      orgId,
      userId,
      token,
      logger,
    )

    const result = await axios
      .get(this.lassoBaseUrl, {
        headers: {
          Authorization: `Bearer ${gatekeeperToken.data.jwt}`,
        },
      })
      .catch(() => {
        logger.error("Problem getting Lasso Token")
        // throw new Error("Problem getting Lasso Token. Might be routes or vpn")
      })

    return result.data.access_token
  }

  async getGatekeeperApiToken(orgId, userId, token, logger) {
    const result = await axios
      .get(
        `${this.gatekeeperBaseUrl}/${orgId}/user/${userId}?scope=portal:admin`,
        {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        },
      )
      .catch(() => {
        logger.error("Problem getting Lasso Token")
        // throw new Error("Problem getting Gatekeeper Token, might be expired JWT")
      })

    return result
  }
}
