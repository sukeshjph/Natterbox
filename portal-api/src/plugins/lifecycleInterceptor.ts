export default {
  // Fires whenever a GraphQL request is received from a client.
  requestDidStart(requestContext) {
    // Do not log introspection as it creates clutter
    if (requestContext.request.operationName !== "IntrospectionQuery") {
      requestContext.context.logger.debug(`Request started`, {
        meta: `${requestContext.request.query}`,
      })
    }

    return {
      willSendResponse({ response }) {
        if (response.errors)
          requestContext.context.logger.error("Error", {
            meta: response.errors,
          })
      },
    }
  },
}
