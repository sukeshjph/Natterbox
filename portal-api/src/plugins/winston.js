import winston from "winston"

const date = new Date().toISOString()

const logFormat = winston.format.printf(info => {
  return `${date} [${info.level}] : ${info.message}\n  ${JSON.stringify(
    info.meta,
    null,
    4,
  ) || ""}`
})

export default winston.createLogger({
  level: process.env.LOG_LEVEL,
  format: (function format() {
    switch (process.env.LOG_FORMAT) {
      case "json":
        return winston.format.json()
      case "text":
        return winston.format.combine(
          winston.format.colorize(),
          winston.format.simple(),
          logFormat,
        )
      default:
        return winston.format.json()
    }
  })(),
  transports: [
    new winston.transports.Console(),
    new winston.transports.File({ filename: "log" }),
  ],
})
