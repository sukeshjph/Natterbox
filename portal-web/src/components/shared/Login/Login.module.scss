@import "../../../styles/core-styles.scss";
