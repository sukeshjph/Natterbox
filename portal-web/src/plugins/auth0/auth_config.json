{
  "domain": "redmatter-qa01.eu.auth0.com",
  "clientId": "sbi7zg4cJX6N8Io0tLtb6dBpom8YBXnq",
  "audience": "https://sapien-proxy.redmatter-qa01.pub/"
}
