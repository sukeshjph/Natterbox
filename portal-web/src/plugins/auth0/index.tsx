/* eslint-disable camelcase */
import React, { useState, useEffect, useContext } from "react"
import createAuth0Client from "@auth0/auth0-spa-js"
import Auth0Client from "@auth0/auth0-spa-js/dist/typings/Auth0Client"

export interface Auth0RedirectState {
  targetUrl?: string
}

export interface Auth0User extends Omit<IdToken, "__raw"> {}

interface Auth0Context {
  user?: Auth0User
  userToken?: string
  isAuthenticated: boolean
  isInitializing: boolean
  isPopupOpen: boolean
  apiUrl: string
  sapienUrl: string
  loginWithPopup(o?: PopupLoginOptions): Promise<void>
  handleRedirectCallback(): Promise<RedirectLoginResult>
  getIdTokenClaims(o?: getIdTokenClaimsOptions): Promise<IdToken>
  loginWithRedirect(o?: RedirectLoginOptions): Promise<void>
  getTokenSilently(o?: GetTokenSilentlyOptions): Promise<string | undefined>
  getTokenWithPopup(o?: GetTokenWithPopupOptions): Promise<string | undefined>
  logout(o?: LogoutOptions): void
}
interface Auth0ProviderOptions {
  children: React.ReactElement
  onRedirectCallback(result: RedirectLoginResult): void
}

export const Auth0Context = React.createContext<Auth0Context | null>(null)
export const useAuth0 = () => useContext(Auth0Context)!

const getAppendedSapienUrl = url => {
  if (!url.includes("v1")) {
    return url.slice(url.length - 1) === "/" ? `${url}v1/` : `${url}/v1/`
  }
  return url
}

const getConfigResponse = async () => {
  if (process.env.NODE_ENV === "development") {
    return {
      graphqlUrl: process.env.REACT_APP_API_URL,
      domain: process.env.REACT_APP_AUTH0_DOMAIN,
      client_id: process.env.REACT_APP_AUTH0_CLIENTID,
      audience: process.env.REACT_APP_AUTH0_AUDIENCE,
      sapien_url: process.env.REACT_APP_SAPIEN_URL,
    }
  }

  let response
  try {
    response = await fetch(`${window.location.origin}/config`)
  } catch (error) {
    throw new Error(
      `Failed to fetch config: Status:${response.status}, message:${error.message}`,
    )
  }

  const configData = await response.json()
  const { authDomain, clientId, sapienUrl, ...others } = configData

  return {
    ...others,
    domain: authDomain,
    client_id: clientId,
    sapien_url: getAppendedSapienUrl(sapienUrl),
  }
}

export const Auth0Provider = ({
  children,
  onRedirectCallback,
  ...initOptions
}: Auth0ProviderOptions & Pick<Auth0ClientOptions, "redirect_uri">) => {
  const [isAuthenticated, setIsAuthenticated] = useState(false)
  const [isInitializing, setIsInitializing] = useState(true)
  const [isPopupOpen, setIsPopupOpen] = useState(false)
  const [user, setUser] = useState<Auth0User>()
  const [userToken, setUserToken] = useState<"">()
  const [auth0Client, setAuth0Client] = useState<Auth0Client>()
  const [apiUrl, setApiUrl] = useState("")
  const [sapienUrl, setSapienUrl] = useState("")

  useEffect(() => {
    const initAuth0 = async () => {
      const {
        audience,
        domain,
        client_id,
        graphqlUrl,
        sapien_url,
      } = await getConfigResponse()

      const auth0FromHook = await createAuth0Client({
        ...initOptions,
        domain,
        client_id,
        audience,
      })

      setAuth0Client(auth0FromHook)
      setApiUrl(graphqlUrl)
      setSapienUrl(sapien_url)

      if (window.location.search.includes("code=")) {
        let appState: RedirectLoginResult = {}
        try {
          ;({ appState } = await auth0FromHook.handleRedirectCallback())
        } finally {
          onRedirectCallback(appState)
        }
      }

      const authed = await auth0FromHook.isAuthenticated()

      if (authed) {
        const userProfile = await auth0FromHook.getUser()
        const token = await auth0FromHook.getTokenSilently()

        setIsAuthenticated(true)
        setUser(userProfile)
        setUserToken(token)
      }

      setIsInitializing(false)
    }

    initAuth0()
  }, [])

  const loginWithPopup = async (options?: PopupLoginOptions) => {
    setIsPopupOpen(true)

    try {
      await auth0Client!.loginWithPopup(options)
    } catch (error) {
      console.error(error) //eslint-disable-line
    } finally {
      setIsPopupOpen(false)
    }

    const userProfile = await auth0Client!.getUser()
    setUser(userProfile)
    setIsAuthenticated(true)
  }

  const handleRedirectCallback = async () => {
    setIsInitializing(true)

    const result = await auth0Client!.handleRedirectCallback()
    const userProfile = await auth0Client!.getUser()

    setIsInitializing(false)
    setIsAuthenticated(true)
    setUser(userProfile)

    return result
  }

  const loginWithRedirect = (options?: RedirectLoginOptions) =>
    auth0Client!.loginWithRedirect(options)

  const getTokenSilently = (options?: GetTokenSilentlyOptions) =>
    auth0Client!.getTokenSilently(options)

  const logout = (options?: LogoutOptions) => auth0Client!.logout(options)

  const getIdTokenClaims = (options?: getIdTokenClaimsOptions) =>
    auth0Client!.getIdTokenClaims(options)

  const getTokenWithPopup = (options?: GetTokenWithPopupOptions) =>
    auth0Client!.getTokenWithPopup(options)

  return (
    <Auth0Context.Provider
      value={{
        user,
        isAuthenticated,
        isInitializing,
        isPopupOpen,
        userToken,
        loginWithPopup,
        loginWithRedirect,
        logout,
        getTokenSilently,
        handleRedirectCallback,
        getIdTokenClaims,
        getTokenWithPopup,
        apiUrl,
        sapienUrl,
      }}>
      {children}
    </Auth0Context.Provider>
  )
}
