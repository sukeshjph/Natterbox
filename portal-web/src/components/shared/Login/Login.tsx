import React from "react"
import IconButton from "@material-ui/core/IconButton"
import { mdiLogout } from "@mdi/js"
import Icon from "@mdi/react"
import { ApolloConsumer } from "@apollo/react-hooks"
import { useAuth0 } from "../../../plugins/auth0"

export const Login = () => {
  const { isAuthenticated, logout } = useAuth0()

  return (
    <ApolloConsumer>
      {client => {
        return (
          <div>
            {isAuthenticated && (
              <IconButton
                onClick={() => {
                  client.resetStore()
                  logout()
                }}>
                <Icon
                  path={mdiLogout}
                  size={1}
                  horizontal
                  color="white"
                  rotate={180}
                />
              </IconButton>
            )}
          </div>
        )
      }}
    </ApolloConsumer>
  )
}
