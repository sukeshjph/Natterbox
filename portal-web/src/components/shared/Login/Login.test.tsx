import React from "react"
import { shallow } from "enzyme"
import toJson from "enzyme-to-json"
import { Login } from "./Login"

import { useAuth0 } from "../../../plugins/auth0"

jest.mock("../../../plugins/auth0")

describe("Login", () => {
  useAuth0.mockReturnValueOnce({
    isAuthenticated: false,
    logout: () => undefined,
  })
  const wrapper = shallow(<Login />)
  it("Renders Login", () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
