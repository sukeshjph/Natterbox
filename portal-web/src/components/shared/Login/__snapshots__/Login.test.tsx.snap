// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Renders Login 1`] = `
<ApolloConsumer>
  <Component />
</ApolloConsumer>
`;
